### PR TITLE
is0501: add testing of SDP files by calling out to sdpoker

### PR DESCRIPTION
**Pull request by andrewbonney**
I'd suggest this may need testing against a few implementations before merging. I'm also not certain how well it will handle non-2110 SDP files, but that may be something we need to contribute back to sdpoker for in order to have a suitable command line argument.

### CODE REVIEW

File: IS0501Test.py [hunk 1]
```diff
@@ -802,6 +803,40 @@ def test_40(self, test):
...
+
+        api = self.apis[CONN_API_KEY]
+        rtp_senders = []
+        if self.is05_utils.compare_api_version(api["version"], "v1.1") >= 0:
+            # Find all RTP senders for v1.1+
+            for sender in self.senders:
+                url = "single/senders/{}/transporttype".format(sender)
+                valid, response = self.is05_utils.checkCleanRequestJSON("GET", url)
+                if valid:
+                    if response == "urn:x-nmos:transport:rtp":
```
Comment (garethsb-sony) [inline IS0501Test.py]: Will need to be adjusted based on final conclusion of https://github.com/AMWA-TV/nmos-device-connection-management/issues/57

File: IS0501Test.py [hunk 2]
```diff
@@ -802,6 +803,40 @@ def test_40(self, test):
...
+            # RTP is the only transport type for v1.0
+            rtp_senders = self.senders
+
+        if len(rtp_senders) == 0:
+            return test.UNCLEAR("Not tested. No resources found.")
+
+        for sender in rtp_senders:
+            path = "single/senders/{}/transportfile".format(sender)
+            try:
+                subprocess.check_output(["sdpoker", "--nmos", "false", self.url + path],
```
Comment (garethsb-sony) [inline IS0501Test.py]: Should the `--shaping true` command line arg be added to check for presence of "TP" format parameter required by ST 2110-21 Section 8.1?

Comment (garethsb-sony) [inline IS0501Test.py]: Should the `--checkEndings true` arg be added to check for plain CR or LF, not CRLF? Perhaps this should be a `WARNING` not a `FAIL` though, so we might need to run sdpoker twice?

Comment (garethsb-sony) [reply IS0501Test.py]: Equally, `--whitespace true`?

Comment (garethsb-sony) [inline IS0501Test.py]: What about the `--should true` option? I haven't examined all the checks that includes, but did failures in that mode ought to be reported as `OPTIONAL`?

Comment (andrewbonney) [reply IS0501Test.py]: I hadn't added any 2110-specific flags as yet given the potential for plain RTP SDP files to be present. Having said that, as noted in the PR notes I don't know how well sdpoker handles non-2110 files. If it doesn't handle them well then some initial checks for mandatory 2110 parameters may be necessary before analysing each file.

Comment (garethsb-sony) [reply IS0501Test.py]: > given the potential for plain RTP SDP files to be present

Oh, OK. Is that true for many IS-04/IS-05 implementations, do you think?

Comment (andrewbonney) [reply IS0501Test.py]: It's certainly true for ours, and I'd hope it'd become increasingly so elsewhere.

**Comment by andrewbonney** [comment]
Added a second run covering some optional checks, and a test for DUP parameters in relevant SDP files. I haven't added '--checkEndings' as it seems to result in an error message of 'undefined'. I've also noted that it's unhappy with non-2110 media types ("For stream 1, encoding name must be media sub-type 'raw', as per SMPTE ST 2110-20 Section 7.1"), so some pre-parsing may be necessary to eliminate those, or a further option adding to sdpoker.

**Comment by garethsb-sony** [comment]
> I haven't added '--checkEndings' 

I think the test itself is good. I submitted a PR to fix the error message itself the week before last... https://github.com/Streampunk/sdpoker/pull/10

**Comment by andrewbonney** [comment]
Ah fair enough, I'll put that back in.

**Comment by garethsb-sony** [comment]
Regarding https://github.com/AMWA-TV/nmos-testing/pull/195#discussion_r274466659, a quick survey of implementations encountered recently shows several not putting in the ST 2110-21 "TP" (type parameter) that is tested for by the `--shaping true` command line option. If ST 2110-21 is viewed as required, then adding the flag is urgent; if ST 2110-21 is optional, probably it should only be added to the second pass if at all?

**Comment by andrewbonney** [comment]
-21 is certainly mandated by the standard when implementing -20. The challenge is likely to be in checking that we're dealing with a -20 SDP and not -30/-40/other as I don't believe sdpoker does this for us after a basic review of the shaping code.

Regarding my more general SDP testing comment, sdpoker doesn't have an option to just check for general SDP errors at present (see https://github.com/Streampunk/sdpoker/blob/master/sdpoker.js#L82) so it would be necessary to detect that we're dealing with a non-2110 SDP and skip testing of those files, or contribute a new command line flag to sdpoker to assist.

**Comment by andrewbonney** [comment]
> I don't believe sdpoker does this for us after a basic review of the shaping code.

Seems that wasn't correct. Looking further into https://github.com/Streampunk/sdpoker/blob/master/checkST2110.js#L460 it does ensure it's only checking for video. What it doesn't appear to do is allow checking of other video codecs.

Based upon that adding '--shaping true' to the first test iteration seems fine for now and will catch some ST.2110 specific stuff, but I suspect we'll hit some issues as further parts of the 2110 suite are released (and with anyone using non-2110 RTP) which may need addressing in sdpoker.

**Comment by garethsb-sony** [comment]
> -21 is certainly mandated by the standard when implementing -20

Do you have a reference for that? I'd like to pass that on...

**Comment by andrewbonney** [comment]
Sure, 4th paragraph of section 6.1.1 ST.2110-20

**Comment by garethsb-sony** [comment]
Added a couple of commits.

One problem I've identified is that if the Sender is disabled (`"master_enable": false`) it should return a 404 for its `/transportfile`. SDPoker doesn't handle that well, it gets an exception which it just logs to console (see below) and doesn't set the return code (so it ends up as zero?) so we get a green 'Pass' for the test.

If all of the discovered Senders are disabled we should probably report UNCLEAR? Or enable them before testing?

We might also want to tweak sdpoker to return non-zero in the case of an exception?
And consider whether output starting 'Found' is enough to distinguish FAIL from DISABLED?

Here's the beginning of a sample output from sdpoker in the 404 case:
```
{ StatusCodeError: 404 - "{\"code\":404,\"error\":\"Sender is not currently configured\",\"debug\":null}"
    at new StatusCodeError (...\AppData\Roaming\npm\node_modules\sdpoker\node_modules\request-promise-core\lib\errors.js:32:15)
    at Request.plumbing.callback (...\AppData\Roaming\npm\node_modules\sdpoker\node_modules\request-promise-core\lib\plumbing.js:104:33)
    at Request.RP$callback [as _callback] (...\AppData\Roaming\npm\node_modules\sdpoker\node_modules\request-promise-core\lib\plumbing.js:46:31)
    at Request.self.callback (...\AppData\Roaming\npm\node_modules\sdpoker\node_modules\request\request.js:185:22)
    at Request.emit (events.js:182:13)
    at Request.<anonymous> (...\AppData\Roaming\npm\node_modules\sdpoker\node_modules\request\request.js:1161:10)
    at Request.emit (events.js:182:13)
    at IncomingMessage.<anonymous> (...\AppData\Roaming\npm\node_modules\sdpoker\node_modules\request\request.js:1083:12)
    at Object.onceWrapper (events.js:273:13)
    at IncomingMessage.emit (events.js:187:15)
  name: 'StatusCodeError',
  statusCode: 404,
  message:
   '404 - "{\\"code\\":404,\\"error\\":\\"Sender is not currently configured\\",\\"debug\\":null}"',
  error:
   '{"code":404,"error":"Sender is not currently configured","debug":null}',
  options:
   { url:
      'http://localhost:3215/x-nmos/connection/v1.0/single/senders/b4163553-a0b0-5205-a026-afa5e82d0a5f/transportfile',
...
```

**Comment by garethsb-sony** [comment]
Apart from the above and https://github.com/Streampunk/sdpoker/issues/11, this is looking good.

**Comment by andrewbonney** [comment]
I'll look at adding something to detect the 404. Otherwise I think we ought to be looking at contributions to sdpoker to resolve further issues.

**Comment by andrewbonney** [comment]
I think that'll sort out the 404 (and other HTTP codes), but I think I'm having some trouble as a result of 'shell=True'. I may be able to resolve this by changing the way the command line is passed in to being a single string rather than a list, but I thought I'd mention it first. I'm not sure if this is a Python version or platform issue.

**Comment by andrewbonney** [comment]
> We've lost `dup_params` here?

That was a mistake/experiment, fixed in https://github.com/AMWA-TV/nmos-testing/pull/195/commits/b4282e8e3a72a1acbdc7051da724a41d2a334888

As noted this only works for me without 'shell=True', but if that's necessary I may need to switch (properly) to using the string rather than list form.

**Comment by garethsb-sony** [comment]
Hmm, I don't know. I think `shell=True` is needed on Windows in order to have PATH searched for 'sdpoker'.

The relationship between using `shell=True` and string vs sequence form is a bit confusing for me. :-/
The docs for `check_output` refer, via `run`, to those for [`Popen`](https://docs.python.org/3/library/subprocess.html#subprocess.Popen)...

**Comment by andrewbonney** [comment]
Ok, based upon that and "If shell is True, it is recommended to pass args as a string rather than as a sequence" from the Python docs I'll change them both to be a string.

**Comment by andrewbonney** [comment]
Yeah, I was hoping I could parse it as JSON and do something a little more specific but it doesn't look like it's _quite_ valid JSON.

**Pull request merged**